File: lib/utils/file-utils.ts
import { generateUUID } from './uuid';

// Utility functions for file operations

export type FileCategory = 'image' | 'video' | 'audio' | 'document' | 'archive' | 'code' | 'folder' | 'other';

export function formatFileSize(bytes: number): string {
    if (bytes === 0) {return '0 B';}

    const k = 1024;
    const sizes = ['B', 'KB', 'MB', 'GB', 'TB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    // Show more precision for smaller files
    const decimals = i <= 1 ? 0 : 1;

    return `${(bytes / Math.pow(k, i)).toFixed(decimals)} ${sizes[i]}`;
}

export function formatSpeed(bytesPerSecond: number): string {
    return formatFileSize(bytesPerSecond) + '/s';
}

export function formatDuration(seconds: number): string {
    if (seconds < 60) {return `${Math.round(seconds)}s`;}
    if (seconds < 3600) {return `${Math.floor(seconds / 60)}m ${Math.round(seconds % 60)}s`;}

    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    return `${hours}h ${minutes}m`;
}

export function getFileIcon(filename: string): string {
    const ext = filename.split('.').pop()?.toLowerCase();

    const iconMap: Record<string, string> = {
        // Images
        jpg: '🖼️',
        jpeg: '🖼️',
        png: '🖼️',
        gif: '🖼️',
        svg: '🖼️',
        webp: '🖼️',

        // Videos
        mp4: '🎬',
        avi: '🎬',
        mov: '🎬',
        mkv: '🎬',
        webm: '🎬',

        // Audio
        mp3: '🎵',
        wav: '🎵',
        flac: '🎵',
        ogg: '🎵',

        // Documents
        pdf: '📄',
        doc: '📝',
        docx: '📝',
        txt: '📝',
        md: '📝',

        // Archives
        zip: '📦',
        rar: '📦',
        '7z': '📦',
        tar: '📦',
        gz: '📦',

        // Code
        js: '💻',
        ts: '💻',
        py: '💻',
        java: '💻',
        cpp: '💻',
        c: '💻',
        html: '💻',
        css: '💻',
    };

    return iconMap[ext || ''] || '📎';
}

/**
 * Get file category from MIME type
 */
export function getFileCategory(mimeType: string): FileCategory {
    if (!mimeType) {return 'other';}

    // Image files
    if (mimeType.startsWith('image/')) {
        return 'image';
    }

    // Video files
    if (mimeType.startsWith('video/')) {
        return 'video';
    }

    // Audio files
    if (mimeType.startsWith('audio/')) {
        return 'audio';
    }

    // Document files
    if (
        mimeType.includes('pdf') ||
        mimeType.includes('document') ||
        mimeType.includes('word') ||
        mimeType.includes('text') ||
        mimeType.includes('msword') ||
        mimeType.includes('officedocument') ||
        mimeType.includes('spreadsheet') ||
        mimeType.includes('presentation')
    ) {
        return 'document';
    }

    // Archive files
    if (
        mimeType.includes('zip') ||
        mimeType.includes('rar') ||
        mimeType.includes('7z') ||
        mimeType.includes('tar') ||
        mimeType.includes('gzip') ||
        mimeType.includes('compress')
    ) {
        return 'archive';
    }

    // Code files
    if (
        mimeType.includes('javascript') ||
        mimeType.includes('typescript') ||
        mimeType.includes('json') ||
        mimeType.includes('xml') ||
        mimeType.includes('html') ||
        mimeType.includes('css') ||
        mimeType.includes('python') ||
        mimeType.includes('java') ||
        mimeType.includes('c++') ||
        mimeType.startsWith('text/x-')
    ) {
        return 'code';
    }

    return 'other';
}

/**
 * Check if file is an image that can be previewed
 */
export function isPreviewableImage(mimeType: string): boolean {
    return (
        mimeType === 'image/jpeg' ||
        mimeType === 'image/jpg' ||
        mimeType === 'image/png' ||
        mimeType === 'image/gif' ||
        mimeType === 'image/webp' ||
        mimeType === 'image/svg+xml'
    );
}

/**
 * Get file extension from filename
 */
export function getFileExtension(filename: string): string {
    const parts = filename.split('.');
    if (parts.length > 1) {
        return parts[parts.length - 1]?.toLowerCase() ?? '';
    }
    return '';
}

export async function fileToArrayBuffer(file: File): Promise<ArrayBuffer> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader();
        reader.onload = (e) => resolve(e.target?.result as ArrayBuffer);
        reader.onerror = reject;
        reader.readAsArrayBuffer(file);
    });
}

export async function createFileChunks(
    file: File,
    chunkSize: number = 1024 * 1024 // 1MB default
): Promise<Blob[]> {
    const chunks: Blob[] = [];
    let offset = 0;

    while (offset < file.size) {
        const end = Math.min(offset + chunkSize, file.size);
        chunks.push(file.slice(offset, end));
        offset = end;
    }

    return chunks;
}

export function generateDeviceId(): string {
    // Generate a persistent device ID
    if (typeof window !== 'undefined') {
        let deviceId = localStorage.getItem('deviceId');
        if (!deviceId) {
            deviceId = generateUUID();
            localStorage.setItem('deviceId', deviceId);
        }
        return deviceId;
    }
    return generateUUID();
}

/**
 * Get platform type using privacy-preserving feature detection
 * Avoids userAgent fingerprinting by using minimal feature detection
 */
export function getPlatform(): 'mobile' | 'desktop' | 'web' {
    if (typeof window === 'undefined') {return 'web';}

    // Use feature detection instead of userAgent parsing
    const isTouchDevice = 'ontouchstart' in window || navigator.maxTouchPoints > 0;
    const isSmallScreen = window.innerWidth < 768;

    // Mobile detection via touch + small screen
    if (isTouchDevice && isSmallScreen) {
        return 'mobile';
    }

    // Desktop (includes tablets in landscape)
    return 'desktop';
}

export function getDeviceName(): string {
    const platform = getPlatform();
    const platformNames: Record<typeof platform, string> = {
        mobile: 'Mobile Device',
        desktop: 'Desktop',
        web: 'Web Browser',
    };

    return platformNames[platform];
}

export async function calculateFileHash(file: File | Blob): Promise<string> {
    const buffer = await file.arrayBuffer();
    const hashBuffer = await crypto.subtle.digest('SHA-256', buffer);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    return hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
}

export function isImageFile(filename: string): boolean {
    const ext = filename.split('.').pop()?.toLowerCase();
    return ['jpg', 'jpeg', 'png', 'gif', 'svg', 'webp', 'bmp'].includes(ext || '');
}

export function isVideoFile(filename: string): boolean {
    const ext = filename.split('.').pop()?.toLowerCase();
    return ['mp4', 'avi', 'mov', 'mkv', 'webm', 'flv', 'wmv'].includes(ext || '');
}

export function downloadBlob(blob: Blob, filename: string) {
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
}

File: tallow-mobile/lib/shared/utils/file_utils.dart
import 'package:flutter/material.dart';

/// File utility functions
class FileUtils {
  /// Get icon for file extension
  static IconData getIconForExtension(String? extension) {
    if (extension == null) return Icons.insert_drive_file;

    switch (extension.toLowerCase()) {
      // Images
      case 'jpg':
      case 'jpeg':
      case 'png':
      case 'gif':
      case 'webp':
      case 'bmp':
      case 'svg':
      case 'heic':
      case 'heif':
        return Icons.image;

      // Videos
      case 'mp4':
      case 'mov':
      case 'avi':
      case 'mkv':
      case 'webm':
      case 'flv':
      case 'wmv':
      case 'm4v':
        return Icons.video_file;

      // Audio
      case 'mp3':
      case 'wav':
      case 'aac':
      case 'flac':
      case 'ogg':
      case 'm4a':
      case 'wma':
        return Icons.audio_file;

      // Documents
      case 'pdf':
        return Icons.picture_as_pdf;
      case 'doc':
      case 'docx':
        return Icons.description;
      case 'xls':
      case 'xlsx':
        return Icons.table_chart;
      case 'ppt':
      case 'pptx':
        return Icons.slideshow;
      case 'txt':
        return Icons.text_snippet;

      // Archives
      case 'zip':
      case 'rar':
      case '7z':
      case 'tar':
      case 'gz':
        return Icons.folder_zip;

      // Code
      case 'html':
      case 'css':
      case 'js':
      case 'ts':
      case 'dart':
      case 'py':
      case 'java':
      case 'kt':
      case 'swift':
      case 'c':
      case 'cpp':
      case 'h':
      case 'json':
      case 'xml':
      case 'yaml':
      case 'yml':
        return Icons.code;

      // APK/Apps
      case 'apk':
      case 'ipa':
      case 'exe':
      case 'dmg':
      case 'deb':
      case 'rpm':
        return Icons.android;

      default:
        return Icons.insert_drive_file;
    }
  }

  /// Get icon for file name
  static IconData getIconForFileName(String fileName) {
    final parts = fileName.split('.');
    if (parts.length > 1) {
      return getIconForExtension(parts.last);
    }
    return Icons.insert_drive_file;
  }

  /// Get file extension from name
  static String? getExtension(String fileName) {
    final parts = fileName.split('.');
    if (parts.length > 1) {
      return parts.last.toLowerCase();
    }
    return null;
  }

  /// Get MIME type for extension
  static String getMimeType(String? extension) {
    if (extension == null) return 'application/octet-stream';

    switch (extension.toLowerCase()) {
      // Images
      case 'jpg':
      case 'jpeg':
        return 'image/jpeg';
      case 'png':
        return 'image/png';
      case 'gif':
        return 'image/gif';
      case 'webp':
        return 'image/webp';
      case 'svg':
        return 'image/svg+xml';
      case 'bmp':
        return 'image/bmp';

      // Videos
      case 'mp4':
        return 'video/mp4';
      case 'webm':
        return 'video/webm';
      case 'mov':
        return 'video/quicktime';
      case 'avi':
        return 'video/x-msvideo';
      case 'mkv':
        return 'video/x-matroska';

      // Audio
      case 'mp3':
        return 'audio/mpeg';
      case 'wav':
        return 'audio/wav';
      case 'ogg':
        return 'audio/ogg';
      case 'flac':
        return 'audio/flac';
      case 'aac':
        return 'audio/aac';
      case 'm4a':
        return 'audio/mp4';

      // Documents
      case 'pdf':
        return 'application/pdf';
      case 'doc':
        return 'application/msword';
      case 'docx':
        return 'application/vnd.openxmlformats-officedocument.wordprocessingml.document';
      case 'xls':
        return 'application/vnd.ms-excel';
      case 'xlsx':
        return 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';
      case 'ppt':
        return 'application/vnd.ms-powerpoint';
      case 'pptx':
        return 'application/vnd.openxmlformats-officedocument.presentationml.presentation';
      case 'txt':
        return 'text/plain';

      // Archives
      case 'zip':
        return 'application/zip';
      case 'rar':
        return 'application/vnd.rar';
      case '7z':
        return 'application/x-7z-compressed';
      case 'tar':
        return 'application/x-tar';
      case 'gz':
        return 'application/gzip';

      // Code
      case 'html':
        return 'text/html';
      case 'css':
        return 'text/css';
      case 'js':
        return 'application/javascript';
      case 'json':
        return 'application/json';
      case 'xml':
        return 'application/xml';

      // Apps
      case 'apk':
        return 'application/vnd.android.package-archive';

      default:
        return 'application/octet-stream';
    }
  }

  /// Check if file is an image
  static bool isImage(String fileName) {
    final ext = getExtension(fileName);
    if (ext == null) return false;
    return ['jpg', 'jpeg', 'png', 'gif', 'webp', 'bmp', 'svg', 'heic', 'heif']
        .contains(ext.toLowerCase());
  }

  /// Check if file is a video
  static bool isVideo(String fileName) {
    final ext = getExtension(fileName);
    if (ext == null) return false;
    return ['mp4', 'mov', 'avi', 'mkv', 'webm', 'flv', 'wmv', 'm4v']
        .contains(ext.toLowerCase());
  }

  /// Check if file is audio
  static bool isAudio(String fileName) {
    final ext = getExtension(fileName);
    if (ext == null) return false;
    return ['mp3', 'wav', 'aac', 'flac', 'ogg', 'm4a', 'wma']
        .contains(ext.toLowerCase());
  }

  /// Check if file is a document
  static bool isDocument(String fileName) {
    final ext = getExtension(fileName);
    if (ext == null) return false;
    return ['pdf', 'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx', 'txt']
        .contains(ext.toLowerCase());
  }

  /// Check if file is an archive
  static bool isArchive(String fileName) {
    final ext = getExtension(fileName);
    if (ext == null) return false;
    return ['zip', 'rar', '7z', 'tar', 'gz'].contains(ext.toLowerCase());
  }

  /// Get file category
  static FileCategory getCategory(String fileName) {
    if (isImage(fileName)) return FileCategory.image;
    if (isVideo(fileName)) return FileCategory.video;
    if (isAudio(fileName)) return FileCategory.audio;
    if (isDocument(fileName)) return FileCategory.document;
    if (isArchive(fileName)) return FileCategory.archive;
    return FileCategory.other;
  }

  /// Sanitize file name for safe storage
  static String sanitizeFileName(String fileName) {
    return fileName
        .replaceAll(RegExp(r'[<>:"/\\|?*]'), '_')
        .replaceAll(RegExp(r'\s+'), '_')
        .replaceAll(RegExp(r'_+'), '_');
  }

  /// Generate unique file name
  static String generateUniqueFileName(String baseName, List<String> existingNames) {
    if (!existingNames.contains(baseName)) return baseName;

    final ext = getExtension(baseName);
    final name = ext != null
        ? baseName.substring(0, baseName.length - ext.length - 1)
        : baseName;

    var counter = 1;
    String newName;
    do {
      newName = ext != null ? '$name ($counter).$ext' : '$name ($counter)';
      counter++;
    } while (existingNames.contains(newName));

    return newName;
  }
}

/// File category enum
enum FileCategory {
  image,
  video,
  audio,
  document,
  archive,
  other,
}
